### PR TITLE
Fix issue #24: Remove all monitoring namespaces

### PR DESCRIPTION
**Pull request by twiest**
Fix issue #24: Remove all monitoring namespaces

- [x] Add label to namespace for easy lookup
- [x] Add annotation to namespace to ensure ownership
- [x] Remove all monitoring projects before run
- [x] Remove all monitoring projects after run
- [x] Only delete projects if they contain our label and annotation
- [x] Test code against cluster
- [x] Test that build still works using: OS_BUILD_ENV_PRESERVE=_output/local/bin hack/env make build-images

### CODE REVIEW

**Comment by twiest** [comment]
/cc @ironcladlou @joelsmith 

I believe this is ready to go. We still need to determine the exact label / annotation, but I'll do that in a separate PR. There is a TODO in the code so I don't forget.

File: cmd/monitor/monitor.go
Comment (ironcladlou) [inline cmd/monitor/monitor.go:244]: This should still return an error to allow the caller to decide how to respond.

File: cmd/monitor/monitor.go
Comment (ironcladlou) [inline cmd/monitor/monitor.go]: Return an error

File: cmd/monitor/monitor.go
Comment (ironcladlou) [inline cmd/monitor/monitor.go]: Try to delete them all, aggregate and return all errors at once.

File: cmd/monitor/monitor.go
Comment (ironcladlou) [inline cmd/monitor/monitor.go]: If this returns an error, we should return early and try again next interval. Otherwise we'll continue to accumulate namespaces in the failure condition the patch is meant to address.

File: cmd/monitor/monitor.go
Comment (ironcladlou) [inline cmd/monitor/monitor.go]: Restore error handling

**Review by ironcladlou — CHANGES_REQUESTED** [review]

**Comment by twiest** [comment]
@ironcladlou I believe I've addressed all of your comments.

Please review to ensure I understood you.

Thanks!

File: cmd/monitor/monitor.go
Comment (ironcladlou) [inline cmd/monitor/monitor.go]: I'm sorry, I forgot that you wouldn't know this without working with Kube code a lot... use k8s.io/apimachinery/pkg/api/errors.NewAggregate to wrap your `[]error` into just an `error`.

File: cmd/monitor/monitor.go
Comment (ironcladlou) [inline cmd/monitor/monitor.go]: Once you're using `Aggregate` you won't need this.

**Review by ironcladlou — CHANGES_REQUESTED** [review]
I suggest these minor error handling refactorings

File: cmd/monitor/monitor.go
Comment (ironcladlou) [inline cmd/monitor/monitor.go]: When using the `Aggregate` error type you won't need this, as the String() implementation will format the error to include all the messages

File: cmd/monitor/monitor.go
Comment (ironcladlou) [inline cmd/monitor/monitor.go]: I think this should be like:

```go
glog.Errorf("failed to clean up projects prior to test: %s", err)`
```

File: cmd/monitor/monitor.go
Comment (ironcladlou) [inline cmd/monitor/monitor.go]: As above, can remove the range and use something like:

```go
glog.Errorf("failed to clean up project following test: %s", err)`
```

**Review by ironcladlou — CHANGES_REQUESTED** [review]
Some additional error handling changes are necessary.

File: cmd/monitor/monitor.go
Comment (twiest) [reply cmd/monitor/monitor.go:244]: I believe I've fixed this one. This comment isn't going away because it's on the removal instead of that addition.

**Comment by twiest** [comment]
@ironcladlou I believe I've addressed your latest feedback (use k8s Aggregate).

Please review again. Thanks!

File: cmd/monitor/monitor.go
Comment (ironcladlou) [inline cmd/monitor/monitor.go:102]: It just occurred to me that we should probably be using backoff here... noting so we don't forget

Comment (ironcladlou) [reply cmd/monitor/monitor.go:102]: Well not here, but instead of `wait.Until`

File: cmd/monitor/monitor.go
Comment (ironcladlou) [inline cmd/monitor/monitor.go]: log the name of the project that was deleted

File: cmd/monitor/monitor.go
Comment (ironcladlou) [inline cmd/monitor/monitor.go]: return `error` instead of `Aggregate`

File: cmd/monitor/monitor.go
Comment (ironcladlou) [inline cmd/monitor/monitor.go]: nit: how about naming this variable `errors`

**Review by ironcladlou — CHANGES_REQUESTED** [review]
Just a few totally minor nits then I think it's good, thanks!

File: cmd/monitor/monitor.go
Comment (twiest) [reply cmd/monitor/monitor.go:102]: I added a TODO comment to implement a backoff. I get the impression that you're ok with that being a future PR. Let me know if I misunderstood your intentions here.

**Comment by twiest** [comment]
@ironcladlou I believe I've addressed all of your comments. Please re-review. Thanks!

**Comment by ironcladlou** [comment]
Did you test a build with:

```
OS_BUILD_ENV_PRESERVE=_output/local/bin hack/env make build-images
```

**Comment by twiest** [comment]
I didn't. Sorry, doing that now.

**Comment by twiest** [comment]
@ironcladlou image build works:

```bash
$ OS_BUILD_ENV_PRESERVE=_output/local/bin hack/env make build-images

[... snip ...]

[openshift/openshift-monitor-project-lifecycle] --> Committing changes to openshift/openshift-monitor-project-lifecycle:36e6d6a ...
[openshift/openshift-monitor-project-lifecycle] --> Tagged as openshift/openshift-monitor-project-lifecycle:latest
[openshift/openshift-monitor-project-lifecycle] --> Done
[INFO] hack/build-images.sh exited with code 0 after 00h 00m 16s
$
```

**Review by ironcladlou — APPROVED** [review]

**Comment by twiest** [comment]
@stevekuznetsov hey, the bot doesn't appear to be merging this PR because "Needs approved label".

I'm not sure what adds that label.

Do I need to give the bot some perms to this github repo or something?

**Comment by elad661** [comment]
@twiest I've seen this problem before on a different repo. The bot can't "create" the label itself, so you need to create the label manually before you use the approve feature for the first time. After the label is created the bot seems to work okay and will be able to tag PRs with that label. (you can also manually label a PR as "approved" yourself, which would create the label)

**Comment by stevekuznetsov** [comment]
That's not correct, the GitHub API will create the label if it does not exist when the bot first attempts to add it.

**Comment by elad661** [comment]
@stevekuznetsov maybe github changed their api? Because this is pretty much what happened in openshift/cluster-monitoring-operator, the bot ignored the "approved" label until someone manually added it to the PR

**Comment by stevekuznetsov** [comment]
@elad661 Can you link me to the PR where that happened? Neither `openshift/cluster-monitoring-operator` nor `openshift/monitor-project-lifecycle` have been configured in our system to honor `/approve`.

**Comment by stevekuznetsov** [comment]
@ironcladlou try `/approve` now.

**Comment by elad661** [comment]
@stevekuznetsov ooh, okay, that explains it. I guess I was confused. Looks like it's configured to look for  the `approved` label ( https://github.com/openshift/release/blob/master/cluster/ci/config/prow/config.yaml#L57 ), so adding the label to the PR manually caused the PR to be merged, but `/approve` didn't actually work

**Comment by stevekuznetsov** [comment]
Yes, manually adding the labels will also work. I've updated the config.

**Comment by stevekuznetsov** [comment]
Y'all also had configured your `master` branch to not allow the bot to merge... should be fine now. `tide` will merge this on the next pass.

**Comment by twiest** [comment]
@stevekuznetsov Thanks for your efforts on this.

Per your comment, has the "next pass" for `tide` run yet?

This still hasn't merged.

**Comment by twiest** [comment]
@stevekuznetsov bump

**Comment by 0xmichalis** [comment]
@twiest `openshift-merge-robot` is not authorized to merge in this repo; see my comments in https://github.com/openshift/release/pull/787#issuecomment-389242766

**Comment by 0xmichalis** [comment]
In latest logs, I am seeing
```
{"base-sha":"9909bbbbfdad70f009ca64f2383184a75f4f038a","branch":"master","component":"tide","controller":"sync","error":"You're not authorized to push to this branch. Visit https://help.github.com/articles/about-protected-branches/ for more information.","level":"error","msg":"Merge failed: Branch needs to be configured to allow this robot to push.","org":"openshift","pr":26,"repo":"monitor-project-lifecycle","sha":"36e6d6a03456a33f2e505dca56dbfd889bb0209e","time":"2018-05-22T18:48:51Z"}
```

**Comment by twiest** [comment]
@kargakis Thanks for responding. "Team OpenShift Robots" which includes "OpenShift CI Robot" has write access to this repo.

I will absolutely do what we need to get the bot the right access, I just don't know what I need to do.

Please let me know what needs to be done.

Thanks!

**Comment by 0xmichalis** [comment]
Seems like branch protection does not allow the bot to merge. I don't have
access in this repo so I can't look into it. :/

**Comment by twiest** [comment]
@kargakis ah, sorry, I didn't know you didn't have access. I've temporarily granted you admin access to this repo.

I'm willing to make whatever changes are needed, I just don't know what needs to be done.

**Comment by 0xmichalis** [comment]
@twiest done, all I did was:

* go to `Settings` > `Branches`
* click `Edit` on the `master` branch
* In the option `Restrict who can push to this branch` type `openshift-merge-robot` in the search bar, select, the robot
* Click `Save changes`

**Comment by twiest** [comment]
@kargakis Perfect, thanks!

**Pull request merged**
